### ISSUE TITLE
cekit as container image

### ISSUE THREAD

**Issue by vorburger** [cekit/cekit#381]
Given #380 and instead of (or in addition to) #378, couldn't you guys just make this tool itself available as a container image? Then someone like me who wants to just run it could use it with something like:

    docker run -it --rm -v $(pwd):/some-dir-in-the-container cekit/cekit

**Comment by vorburger** [comment]
Such an image, if it was even possible to create, should probably include frequently used tools, e.g. `python3-behave` and whatever `lxml` is (see https://github.com/jboss-container-images/openjdk/issues/63) .. 

I'm starting to understand that cekit, contrary e.g. to https://github.com/fabric8io-images/fish-pepper, is not just a template thingie, but actually wants to drive docker itself; this may not be trivial from within docker ("docker-in-docker") ... although for buildah it should be possible, I guess.

**Comment by rnc** [comment]
I have actually had a look at this (as I only have podman installed) but have come to the conclusion that its going to be rather difficult to get docker in docker (so to speak) actually working. It may be easier to provide a VM (vagrantfile maybe / ansible configuration / etc. However that is more for development than actually using it.

**Comment by jmtd** [comment]
`generate` itself should be containerizable without docker-in-docker problems, at least (leaving you a build tree you could pass to `docker` or `buildah`). We actually played with this as the default method of using it some time ago, when the tool was called dogen: https://github.com/jboss-dockerfiles/dogen/blame/develop/README.rst#L31

I'm not sure why we moved away from that when we evolved the tool to cekit, @goldmann might have an idea

**Comment by jmtd** [comment]
> Such an image, if it was even possible to create, should probably include frequently used tools, e.g. `python3-behave` and whatever `lxml` is (see [jboss-container-images/openjdk#63](https://github.com/jboss-container-images/openjdk/issues/63)) ..

Since a lot of the python dependencies are only such for particular combinations of cekit features: that is, lxml, behave etc only if you run tests; docker-py only if you use the docker build back-end, etc., the current maintainers decided to put only the minimal always-needed dependencies into the requirements file (and the RPM dependencies).

I wonder if we could include a "most likely needed for most people" set of dependencies somewhere else; perhaps even a virtual RPM package that depended on the stuff you are likely to need if you are building jboss middleware openshift images. I have the exact same issue whenever I upgrade my cekit installation.

**Comment by goldmann** [comment]
TL;DR; We were there, it didn't fly. Don't want to go through it again.

To make it possible to run Docker inside Docker you need to run the container as a privileged container and expose the host's Docker inside of the container. Running true Docker in a container will be problematic and not sure if even doable.

Other issue you will face later are permissions. Files generated by Cekit inside of the container will have different uid/gid so every time you will need to chown these to the correct state. Figuring out what it should be can be hacky.

As @jmtd already pointed out, we already used this in a container, it proved to be error prone and time we spend to maintain the container was not worth it. We decided that better way to provide the tooling will be a RPM repository and this is what we have today. And it works great. Other issues you may imagine is - how to forward Kerberos to a container? How about configuration files and multiple directories (target, dist-git, cache) - are you sure you will know from where and to where mount all of them? And so on...

Regarding dependencies, I already commented on it on several places, one of them: https://github.com/cekit/cekit/issues/380#issuecomment-463131387. This has links to other places. Try `python3-cekit-next` package.

Considering that we already moved out of the idea of containerizing similar tooling I don't want to go that route again because it was a pain.

**Comment by jmtd** [comment]
@goldmann 
> Regarding dependencies, I already commented on it on several places, one of them: [#380 (comment)](https://github.com/cekit/cekit/issues/380#issuecomment-463131387). This has links to other places. Try `python3-cekit-next` package.

What do you think about the idea of a new RPM meta-package that pulled in maximal dependencies as an optional alternative to the default cekit one, or at least those that would be exercised by the features that are going to be used by jboss middleware people (buildah, fedpkg, rhpkg, odcs-client, lxml, behave, (does the testing still depend upon docker-py specifically?), s2i (is that packaged in fedora?) etc…)

**Comment by goldmann** [comment]
The meta-package would not be possible to define easily and the reason is that dependencies come from different repositories (not available to many people - example: `rhpkg`), or there are no packaged dependencies (example is the `s2i` above). My decision was to go different route - go minimal and fail in the most understandable way pointing out what could be needed. I think this is better that to try to solve everything and then fail in successfully delivering the promise.
